Sources/_MatchingEngine/Regex/Parse/SourceLocation.swift

```diff
@@ -0,0 +1,135 @@
+extension Source {
+  /// The location in the input of a parsed entity, presented as a region over the input
+  public struct Location: Hashable {
+    public var range: Range<Source.Position>
+
+    public var start: Source.Position { range.lowerBound }
+
+    /// The open end
+    public var end: Source.Position { range.upperBound }
+
+    public init(_ r: Range<Source.Position>) {
+      self.range = r
+    }
+    public init<R: RangeExpression>(
+      _ r: R, in input: Source
+    ) where R.Bound == Source.Position {
+      self.init(r.relative(to: input.input))
+    }
+
+    /// A "fake" location, sometimes useful for tooling purposes
+    public static var fake: Self {
+      .init("".startIndex ..< "".startIndex)
+    }
+    public var isFake: Bool { self == Self.fake }
+    public var isReal: Bool { !isFake }
+  }
+}
+public typealias SourceLocation = Source.Location
+
+extension Source {
+  var currentPosition: Position { bounds.lowerBound }
+}
+
+extension Source {
+  /// An error with source location info
+  public struct Err<E: Error>: Error {
```

Comment:
Could we use a more descriptive name instead of an abbreviation that could be easily misinterpreted? Suggest `LocatedError<E>`.

Reply:
Sounds good, though I think we're likely going to make this internal and formally use a ParseError struct with the location. These wrappers are trending towards being useful for lexical analysis and holding sourced/located values, as opposed to AST nodes themselves. Diagnostics will likely follow suit.